csharp/src/Ice/IceDiscovery/LocatorRegistry.cs

```diff
@@ -0,0 +1,298 @@
+// Copyright (c) ZeroC, Inc. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+
+using ZeroC.Ice;
+
+namespace ZeroC.IceDiscovery
+{
+    /// <summary>Servant class that implements the Slice interface Ice::LocatorRegistry.</summary>
+    internal class LocatorRegistry : ILocatorRegistry
+    {
+        private readonly IObjectPrx _dummyIce1Proxy;
+        private readonly IObjectPrx _dummyIce2Proxy;
+
+        private readonly Dictionary<string, IObjectPrx> _ice1Adapters = new ();
+        private readonly Dictionary<string, IReadOnlyList<EndpointData>> _ice2Adapters = new ();
+
+        private readonly object _mutex = new ();
+
+        private readonly Dictionary<(string, Protocol), HashSet<string>> _replicaGroups = new ();
+
+        public void RegisterAdapterEndpoints(
+            string adapterId,
+            string replicaGroupId,
+            EndpointData[] endpoints,
+            Current current,
+            CancellationToken cancel)
+        {
+            if (adapterId.Length == 0)
+            {
+                throw new InvalidArgumentException("adapterId cannot be empty", nameof(adapterId));
+            }
+            if (endpoints.Length == 0)
+            {
+                throw new InvalidArgumentException("endpoints cannot be empty", nameof(endpoints));
+            }
+
+            lock (_mutex)
+            {
+                try
+                {
+                    _ice2Adapters.Add(adapterId, endpoints);
+                }
+                catch (ArgumentException)
+                {
+                    throw new AdapterAlreadyActiveException($"adapter `{adapterId}' is already active");
```

Comment:
Won't this prevent the server to register again its endpoints after a crash?

Reply:
The Locator and LocatorRegistry are colocated with the server's OAs, so what's your concern?

Reply:
Ok, I forgot about that. Should this be an assert then? And can't   actually be a problem if the OA published endpoints are refreshed however? We should fix  the OA impl to first unregister the endpoints or we do already?

Reply:
Fixed IceDiscovery to always replace the endpoints and not throw AdapterAlreadyActiveException, just like IceDiscovery in Ice 3.7.